more H2 contrib mappings

ndushay opened the pull request:
## Why was this change made?

Part of #1375 - ensuring that all the mappings needed for contributors by h2 are coded here.

Mapping changes in this PR 
- Cocina contributor type `event` added but it is mapped to `corporate` in MODS, which already has a mapping to Cocina `organization` (in other words, it's not a reciprocal mapping)
- accommodates a cocina contributor having multiple roles.
- DataCite event data should not become a cocina `form`

There are 3 failing specs but I want to put those in a separate PR to avoid potential merging problems, as there is an open PR on `originInfo` mappings, which are implicated in these failing specs.

## How was this change tested?

I used https://github.com/sul-dlss-labs/cocina-descriptive-metadata/blob/master/h2_cocina_mappings/h2_to_cocina_contributor.txt as my source for the two new spec files added;  I was glad for all the existing specs so I knew whether my changes were messing up anything else.

## Which documentation and/or configurations were updated?

ndushay commented on app/services/cocina/to_fedora/descriptive/contributor.rb:
```diff
@@ -5,7 +5,10 @@ module ToFedora
       # Maps contributors from cocina to MODS XML
       class Contributor
-        NAME_TYPE = FromFedora::Descriptive::Contributor::ROLES.invert.freeze
+        name_types = FromFedora::Descriptive::Contributor::ROLES.invert
+        # one way mapping:  MODS 'corporate' already maps to Cocina 'organization'
+        name_types['event'] = 'corporate'
+        NAME_TYPE = name_types.freeze
```
this is another funky thing, but there's another class that uses NAME_TYPE as a class variable, so ... my attempts to refactor this died.   Again, would rather have the working code merged and deal with the refactors after ...

mjgiarlo commented on app/services/cocina/to_fedora/descriptive/contributor.rb:
```diff
@@ -5,7 +5,10 @@ module ToFedora
-        NAME_TYPE = FromFedora::Descriptive::Contributor::ROLES.invert.freeze
+        name_types = FromFedora::Descriptive::Contributor::ROLES.invert
+        # one way mapping:  MODS 'corporate' already maps to Cocina 'organization'
+        name_types['event'] = 'corporate'
```
So we want to modify the `FromFedora::Descriptive::Contributor::ROLES` constant for use in the `ToFedora::Descriptive::Contributor` class but not in the FromFedora one?

ndushay replied:
yes - that is what I mean by "one way mapping" -- the FROM hash would have one key going to two different values.  This is nasty ... but it does work.  I tried to clarify with the comment.

mjgiarlo replied:
OK. would you consider this small tweak to what you have?

```ruby
NAME_TYPE = Cocina::FromFedora::Descriptive::Contributor::ROLES.invert.merge('event' => 'corporate').freeze
```

I think the comment is OK.

ndushay replied:
hahaha - i couldn't for the life of me remember the `.merge` thing and was doing contortions and it seemed so wrong ....

ndushay commented on app/services/cocina/to_fedora/descriptive/form.rb:
```diff
@@ -69,6 +69,7 @@ def write_physical_description(forms)
         end
 
         def write_basic(form)
+          return nil if form.source&.value&.match?(/DataCite/i)
```
yes, this is needed from h2 work.  🤮

ndushay commented on spec/services/cocina/from_fedora/descriptive_spec.rb:
```diff
@@ -301,42 +301,48 @@
-              code: 'marcrelator',
-              uri: 'http://id.loc.gov/vocabulary/relators/'
+          role: [
+            { value: 'degree supervisor' },
+            {
```
here the multiple roles for a single contributor affect pre-existing specs ...

mjgiarlo commented on app/services/cocina/to_fedora/descriptive/contributor.rb:
```diff
@@ -56,21 +59,41 @@ def name_attributes(contributor)
+          mr_roles_xml = marcrelator_roles_xml(contributor)
+          return mr_roles_xml if mr_roles_xml.present?
+
+          Array(contributor.role).each { |role| xml_role(role) unless role.value&.match?(/conference/i) }
```
Do you intend to return the whole `Array(contributor.role)` here or a separate array of xml_role info where role.value doesn't match conference? I thought the return value of `#each` was the array it was invoked on. So maybe you want to return `Array(contributor.role)` but you also want to cause side effects? I'm a bit confused on this one.

ndushay replied:
I want an Array, that is comprised of contributor roles, to call xml_role for each role EXCEPT one case.  I will try it right now with just "contributor.role" and see if that works.

mjgiarlo commented on app/services/cocina/to_fedora/descriptive/contributor.rb:
```diff
@@ -56,21 +59,41 @@ def name_attributes(contributor)
+
+        def marcrelator_roles_xml(contributor)
+          result = []
+          contributor.role.select do |role|
```
I'm a bit puzzled by using `#select` with a block that doesn't seem to have a condition, but that does an append side effect.

ndushay replied:
i think this evolved as I wrote it.  I'll do an each.

mjgiarlo commented on app/services/cocina/from_fedora/descriptive/contributor.rb:
```diff
@@ -143,44 +144,59 @@ def names
+        ROLE_AUTHORITY_VALUE_XPATH = './mods:roleTerm/@valueURI'
+        MARC_RELATOR_PIECE = 'id.loc.gov/vocabulary/relators'
+
+        # shameless green
```
I know you will follow this up in a forthcoming PR, so no worries.

mjgiarlo reviewed (approved):
Please test against a sufficiently large dataset in prod (sdr-deploy) and then feel free to merge if the results look good.

ndushay commented:
oh my.  very slightly worse.  what do I do?

### Validations Before:
(roundtrip

```
Status (n=500000):
  Success:   182906 (36.5812%)
  Different: 313852 (62.7704%)
  To Cocina error:     203 (0.0406%)
  To Fedora error:     85 (0.017%)
  Missing:     2954 (0.5908%)
```

### Validations After
```
Status (n=500000):
  Success:   181909 (36.3818%)
  Different: 314919 (62.9838%)
  To Cocina error:     206 (0.0412%)
  To Fedora error:     12 (0.0024%)
  Missing:     2954 (0.5908%)
```

justinlittman commented:
I've gotten bit before by not being rebased off the latest master, which throws off the comparison.

ndushay commented:
thanks, Justin.  I did notice (before I saw your comment) that I wasn't testing my branch against the same master, so I fixed that and adjusted stats above.

ndushay commented on app/services/cocina/from_fedora/descriptive/contributor.rb:
```diff
@@ -193,37 +209,26 @@ def type_for(type)
             return
           end
 
-          raise Cocina::Mapper::InvalidDescMetadata, "Contributor role code is missing and has unexpected value: #{role_code.content}"
```
this error message didn't make sense, so i tweaked it.
